### DISCUSSION TITLE
Floating point error

### DISCUSSION THREAD

**Discussion by Shomari-Healy** [EDmodel/ED2#408] [General]
Hi, I am getting this floating-point exception error in the thermlib.f90 when running the model and I am not sure why I am getting this. I have tried adding print statements to file to see what values the variables are taking but I am not getting any results on the command line or in any log files that I can see. I also checked the ed_met_driver.f90, ed_params.f90, and ed_driver.f90 files but I cannot see where the issue is coming from. I then thought the problem might be with my input meteorology files but as far I can tell they are fine (my Met header is below).

1
/rds/homes/s/sch315/changl-regional-simulations/BIFoR_FACE_sim/PV/FACE_weather_hr/FACE_weather_hr_
2 2 1 1 -2.301 52.801
12
   'hgt'   'tmp'  'pres'    'sh'  'ugrd'  'vgrd' 'prate' 'dlwrf' 'nbdsf' 'nddsf' 'vbdsf' 'vddsf'
27 3600 3600 3600 3600 3600 3600 3600 3600 3600 3600 3600
 4 1 1 1 1 1 0 1 1 1 1 1

I have also uploaded my output file and a screenshot of the error. If you have any advice on how to solve this issue that would be great as well as any advice in general. I can upload any files if you would like to have a look at anything else. 
[output.txt](https://github.com/user-attachments/files/17520669/output.txt)
 
<img width="1710" alt="Screenshot 2024-10-25 at 12 40 52" src="https://github.com/user-attachments/assets/b570e63f-6be4-4aed-9307-c94a5543ad57">

**Comment by mpaiao** [comment]
@Shomari-Healy Sorry, I don't have any good guess, but it looks like the model is crashing even before going through the time steps. This suggests something is off either in the initial conditions or in the meteorological drivers.

Some suggestions:
1. Which NL%IED_INIT_MODE are you using? Unless you are running a near-bare ground simulation (NL%IED_INIT_MODE =0), I suggest configuring the initial conditions (pss/css files) following the [NL%IED_INIT_MODE=6](https://github.com/EDmodel/ED2/wiki/Initial-conditions#files-types-and-formats-for-nlied_init_mode6) standard.
2. Are you trying to run TOPMODEL by any chance? If so, I don't think that part of the code is functional, though it would be great to make it work again.
3. Did you generate the meteorological drivers yourself? If so, I suggest using [these guidelines](https://github.com/EDmodel/ED2/wiki/Drivers#input-meteorology) for generating the drivers, paying special attention to units. Either way, I suggest making sure that the ED2IN variable NL%IMETAVG is set consistently with the definition of the time stamps in the meteorological driver (when I have time, I want to make a pull request moving IMETAVG to be part of the meteorological driver header to ensure this is established by whoever creates the meteorological drivers).

If you share your ED2IN we can try to see if we can spot anything else.

**Comment by Shomari-Healy** [comment]
Hi Marcos,

Thanks for getting back to me. Yes, my guess was the met drivers as well. I
am using NL%IED_INIT_MODE = 0 as I am trying to run a near-bare ground
simulation. I am not sure if I am trying to run TOPMODEL, however. I am
using version 2.2 and just installed it from the GitHub. How do I check if
I am using TOPMODEL accidentally?

And yes, I generated the met drivers myself. I followed the guidelines on
the GitHub and used the R scripts provided to generate my files. I am
pretty sure all of my units are correct but I will double check just to
make sure. My NL%IMETAVG value is set to 1 in my ED2IN but I tried it at 0
as well and got the same result. I have attached it for you to have a look
at, as well as some of my MET files. Please let me know if there is
anything else I can send you. I appreciate the help.

**Comment by mpaiao** [comment]
@Shomari-Healy I did not get any attachment, I think you need to upload them through the GitHub website. 

If you are running the model with near-bare ground initial conditions, then you are not running TOPMODEL. I also doubt that the problem has anything to do with IMETAVG because it is a floating point error happening at the very first time step. I think the best way to figure out this would be for you to recompile the code with very stringent debugging flags and run the code again. By doing this, the code will crash at the very first exception, so we can have a better sense of what is causing the problem.

To recompile the code with strict debugging flags, you can check some include.mk files in directory [ED/build/make/ ](ED/build/make). The best one will depend on which compiler you are using:

- GNU (gcc/gfortran). Check [ED/build/make/include.mk.macosx](ED/build/make/include.mk.macosx)
- Intel (icc/ifort). Check [ED/build/make/include.mk.intel](ED/build/make/include.mk.intel)

In either case, you want to have the compilation settings for the `KIND_COMP=A` option.  Your include.mk file may already have the settings needed for compiling with `KIND_COMP=A`.

**Reply by Shomari-Healy** [reply]
Hi Marcos,

Sorry, I tried to attach them by email but they seemed to have not sent. GitHub won't let me attach them directly to this thread for some reason so I will send them to you personally instead, if that is ok.

That's a good idea thank you. I have been compiling it with these flags (./install.sh -k C -p gfortran) but I will have a look at the mk files directory and add the additional flags. Hopefully I can get a better clue as to what the problem is then.
